## Type: Arrays & Hashing

### LC 217: Contains Duplicate

<img src="img/lc217.png" style="width:600px;height:600px"/>

In [1]:
def containsDuplicate(nums:list[int]) -> bool:
    return len(set(nums)) != len(nums)

def containsDuplicate(nums:list[int]) -> bool:
    # create a set to store the nums
    nums_set = set()

    # iterate over nums
    for num in nums:

        # check if num is in nums_set already, if it is, then return True
        if num in nums_set:
            return True

        # Otherwise add this num to num_set
        nums_set.add(num)

    return False


In [2]:
containsDuplicate([1,4,2,3])

False

In [12]:
containsDuplicate([1,2,2,3])

True

**Time Complexity: O(n), Space Complexity: O(n)**

### LC 242: Valid Anagram

<img src = "img/lc242.png" style="width:600px;height:600px"/>

In [54]:
### Approach 1: Sort the strings and check if they are the same
def isAnagram(s, t):
    s = sorted(s)
    t = sorted(t)
    return s == t
    
### Approach 2: Keep frequency count of all unique characters for one string and compare with the other string
def isAnagram(s, t): 
    char_frequency = {}
    length = len(s)
    
    # base case
    if len(s) != len(t):
        return False
    
    for i in range(length):
        char_s = s[i]
        char_t = t[i]
        
        # increment the count of the char at this index in s
        char_frequency[char_s] = char_frequency[char_s] + 1 if char_s in char_frequency else 1
        
        # decrement the count of the char at this index in t
        char_frequency[char_t] = char_frequency[char_t] - 1 if char_t in char_frequency else -1
        
    for k, v in char_frequency.items():
        if v != 0:
            return False
        
        
    return True

        
    

In [56]:
isAnagram("abc", "acb")

True

### LC 1299: Replace Elements With Greatest Element On Right Side

<img src = "img/lc1299.png" style="width:500px;height:550px"/>

In [17]:
def replaceElements(arr):
    # initialize values
    length   = len(arr) # length
    max_num  = arr[-1] # max elem
    
    # base case
    if length == 0:
        return []
    
    # base case
    if length == 1:
        return [-1]
    
    # replace first elem with -1
    arr[-1] = -1
    
    # iterate array in reverse order
    for i in range(1, length): 
        index = length-i-1 # reverse index Note: -1 added to account for index number calc
        current_num = arr[index]
        
        # replace current num with the max num
        arr[index] = max_num
        
        # compare current num and max num and update 
        max_num = max(max_num, current_num)
    
    return arr
        

In [18]:
replaceElements([17, 18, 5, 4, 6, 1])


[18, 6, 6, 6, 1, -1]

In [19]:
replaceElements([400])

[-1]

### LC 1: Two Sum

<img src = "img/q1.png" style="width:700px;height:700px"/>

In [ ]:
class Solution:
    def twoSum(self, nums:list[int], target: int) -> list[int]:
        hashmap = {}
        
        for i in range(len(nums)):
            
            #compute difference
            diff = target - nums[i]
            
            # check if complement in hashmap
            if diff in hashmap:
                return [hashmap[diff], i]
            
            # else add this num and index to hashmap
            hashmap[nums[i]] = i
        

In [4]:
class Solution:

    def twoSum(self, nums: list[int], target: int) -> list[int]:

        # Initialize a num to index map
        nums_to_index = {}

        # iterate over the nums
        for i, num in enumerate(nums):

            # compute the difference between target and num
            diff = target - num

            #if the diff number is already in the map
            if diff in nums_to_index:
                return [nums_to_index[diff], i] # return the index of the diff and this index

            #Otherwise just add it to the map and continue iterating
            nums_to_index[num] = i


        return

solution = Solution()


In [5]:
solution.twoSum([3,3], 6)

[0, 1]

In [6]:
solution.twoSum([3,7,2,11,15], 9)

[1, 2]

**Time Complexity: O(n), Space Complexity: O(n)**

Time complexity: O(n). We traverse the list containing n elements only once. Each lookup in the table costs only O(1) time.
Space complexity: O(n). The extra space required depends on the number of items stored in the hash table, which stores at most n elements.

### LC 169: Majority Element

<img src = "img/lc169.png" style="width:400px;height:600px"/>

In [8]:
def getMajority(nums):
    nums.sort()
    return nums[len(nums)//2]

In [12]:
def getMajority(nums):
    ## store count and majority number
    count = 0
    maj_elem = nums[0]
    
    #iterate over nums
    for num in nums:
        if count != 0:
            if num == maj_elem:
                count += 1
            else:
                count -= 1
                
        # if count is 0
        if count == 0:
            maj_elem = num
            count = 1
    # return majority elem, which is guarenteed to exist in this case        
    return maj_elem

In [13]:
getMajority([3,2,3])

3

In [14]:
getMajority([2,2,1,1,1,2,2])

2

## Type: Two pointers

### LC 125: Valid Palindrome

<img src = "img/vp.png" style="width:500px;height:500px"/>

In [15]:
def isPalindrome(s:str) -> bool:
    # 1: iterate through string and include only alphanumeric characters,
    # 2: Then join the string and convert to lowercase
    # 3: Then check if reverse of string equals the processed string
    s_alphanum =  "".join([ch for ch in s if ch.isalnum()]).lower()
    return s_alphanum == s_alphanum[::-1]

In [16]:
isPalindrome("A man, a plan, a canal: Panama")

True

In [17]:
isPalindrome("race a car")

False

In [13]:
isPalindrome("")

True

In [91]:
def isPalindrome(s:str) -> bool:
    
    #Approach have a start and end pointers and compare if both pointers correspond to the same character
    
    # Initialize two pointers
    start_ptr, end_ptr = 0, len(s)-1
    
    # Iterate until the pointers meet
    while start_ptr < end_ptr:
        
        # Preprocess: increment or decrement until both pointers have alpha-numeric char (ignore spaces, commas etc)
        while not s[start_ptr].isalnum() and start_ptr < end_ptr:
            start_ptr += 1
        while not s[end_ptr].isalnum() and start_ptr < end_ptr:
            end_ptr -= 1
            
        
        # palindrome is valid if the characters at start and end pointer is same (note: lowercase all chars first)
        if s[start_ptr].lower() != s[end_ptr].lower():
            return False
        
        # increment start ptr
        start_ptr += 1
        # decrement end ptr
        end_ptr   -= 1
        
    return True

In [89]:
isPalindrome("aba")

True

In [90]:
isPalindrome("abc")

False

In [87]:
ss = "ab c"

ss[0].isalnum()

True

### LC 680: Valid Palindrome II

<img src = "img/lc680.png" style="width:450px;height:500px"/>

### LC 283: Move Zeroes

<img src = "img/lc283.png" style="width:470px; height:600px"/>

In [13]:
def moveZeroes(nums):
        """
        Do not return anything, modify nums in-place instead.
        """
        # initialize pointer
        left  = 0
        
        # iterate until start and end overlap
        for right in range(len(nums)):
            
            # if non-zero element
            if nums[right] != 0:
                # swap values between index with 0 and index with non-zero
                nums[left], nums[right] = nums[right], nums[left]
                
                #increment left pointer
                left += 1
                
            # else: if num is 0 just continue iterating
            

In [14]:
nums = [0,1,0,3,12]
moveZeroes(nums)
nums

[1, 3, 12, 0, 0]

In [15]:
nums = [0]
moveZeroes(nums)
nums

[0]

## Type: Sliding Window

## Type: Binary Search

### LC 977: Squares of a Sorted Array
Binary Search

<img src = "img/lc977.png" style="width:400px;height:500px"/>

In [23]:
def sortSquaredArray(nums):
    # initialize left and right pointers as first and last index
    lp, rp = 0, len(nums)-1

    # output = []
    output = [1] * len(nums)

    # iterate until the points don't overlap
    for i in range(len(nums)-1,-1,-1):

    # while lp < rp:
        # start with the tip of the lists, note that the largest terms that exist should be towards the right tip or left tip if any negative number exists,
        left_term_sq  = nums[lp]**2
        right_term_sq = nums[rp]**2


        # whichever term is larger add that to the result and update the corresponding pointer
        if left_term_sq>right_term_sq:
            # output.append(left_term_sq) # append the larger term
            output[i] = left_term_sq
            lp +=1

        else:
            # output.append(right_term_sq) #append the right term
            output[i] = right_term_sq
            rp -= 1 # decrement the right pointer

    # now return the reverse list
    # output.reverse()
    return output

In [24]:
sortSquaredArray([-4,-1,0,3,10])

[0, 1, 9, 16, 100]

In [25]:
sortSquaredArray([-7,-3,2,3,11])

[4, 9, 9, 49, 121]

## Type: Stack

### LC 20: Valid Parentheses

<img src="img/lc_20.png" style="width:500px;height:500px" />

In [21]:
def isValidParentheses(string:str) -> bool:
    # intialise a hashmap of open and closed bracket pairs
    brackets = {"}" : "{", "]" : "[", ")" : "("}
    
    # initialize a stack (list)
    stack = []
    
    ### iterate over the string
    
    for ch in string:
        
        ## 1. if ch is closed bracket
        if ch in brackets:
        
            # 2. if stack is not empty before closed bracket
            if stack:
                ## pop the last open bracket from stack & 
                last_open = stack.pop() 
                
                
           # 2. if stack is empty means closed bracket precedes open bracket which is invalid
            else:
                return False
            
            ## 3. check if the closed bracket corresponds to the open brack (dict of pairs of open close brackets)
            if last_open != brackets[ch]:
                print(last_open, brackets[ch])
                return False ## if incorrect return false
            
        ## 1. else if ch is open bracket, store it to a stack 
        else:
            stack.append(ch)

            
    ## At the end of loop, return True if stack is empty (i.e all open brack matched with closed brack)
    return not stack

In [22]:
isValidParentheses("()")

True

In [23]:
isValidParentheses("()[]{}")

True

In [24]:
isValidParentheses("[]](")

False

#### Complexity analysis

#### Time complexity :
O(n) because we simply traverse the given string one character at a time and push and pop operations on a stack take O(1) time.
#### Space complexity : 
O(n) as we push all opening brackets onto the stack and in the worst case, we will end up pushing all the brackets onto the stack. e.g. ((((((((((.

In [1]:
def hasValidParentheses(s:str) -> bool:
    # initialize a map with closed to open bracket pairs
    pairs = {"}":"{", "]":"[", ")":"("}
    stack = []

    # iterate through string
    for ch in s:

        # if open bracket, add to stack
        if ch not in pairs:
            stack.append(ch)

        # if close bracket, pop the stack to check the last open bracket was the corresponding open bracket for the closed bracket
        else:
            open_bracket = stack.pop()
            # if last open bracket is not the first closed bracket, return False
            if pairs[ch] != open_bracket:
                return False

    # if correct and stack is empty at the end with all pairings popped, its valid
    return False if stack else True


In [2]:
hasValidParentheses("()")

True

In [3]:
hasValidParentheses("()[]{}")

True

In [4]:
hasValidParentheses("(]")

False

## Type: Intervals

<img src = "img/lc252.png" style="width:550px;height:580px"/>

In [3]:
a = [[1,2], [3,1], [2, 3]]

a.sort()
a
#Note: sort() on list of lists sorts by the first index of sublist

[[1, 2], [2, 3], [3, 1]]

In [11]:
class Solution:
    def canAttendMeeting(self, intervals) -> bool:
        
        # sort intervals by start time
        intervals.sort()
        
        # iterate over sorted intervals
        for i in range(len(intervals) - 1): # -1 since we are comparing a pair of intervals per iteration
            # check if end time of current is more than start time of next 
            if intervals[i][1] > intervals[i+1][0]:
                return False
            
            # else continue
            
        # if loop completes return True
        return True
        

In [12]:
soln = Solution()
soln.canAttendMeeting(intervals = [[0,30],[5,10],[15,20]])

False

In [13]:
soln = Solution()
soln.canAttendMeeting(intervals = [[7,10],[2,4]])

True

Time complexity : O(nlog⁡n). The time complexity is dominated by sorting. Once the array has been sorted, only O(n) time is taken to go through the array and determine if there is any overlap.

Space complexity : O(1). Since no additional space is allocated.

In [27]:

"""
Definition of Interval:
"""
class Interval(object):
    def __init__(self, start, end):
        self.start = start
        self.end = end
        
class Solution:
    """
    @param intervals: an array of meeting time intervals
    @return: if a person could attend all meetings
    """
    def can_attend_meetings(self, intervals):
        # sort the intervals by start time
        intervals.sort(key = lambda interval: interval.start)
        
        #iterate over intervals
        for i in range(len(intervals)-1): # -1 since # intervals is less than nums
            
            # compare consecutive pairs and return False
            
            # if the intervals overlap (end time of this is later than start time of other)
            if intervals[i].end > intervals[i+1].start:
                return False
        
        return True

In [28]:
nums = [(5,10),(0,30),(15,20)]
intervals = [Interval(i[0],i[-1]) for i in nums]   

soln = Solution()
soln.can_attend_meetings(intervals)

False

In [29]:
nums = [(5,8),(9,15)]
intervals = [Interval(i[0],i[-1]) for i in nums]   

soln = Solution()
soln.can_attend_meetings(intervals)

True

<img src = "img/lc253.png" style="width:550px;height:550px"/>

<img src = "img/lc252f1.png" style="width:500px;height:450px"/>

<img src = "img/lc252f2.png" style="width:500px;height:400px"/>

In [29]:
def minMeetingRooms(intervals):
    # separate start and end times and sort them
    start = sorted([i[0] for i in intervals]) # sorted for returning the list after sorting
    end   = sorted([i[1] for i in intervals]) 
    
    # initialize two pointers for start and end
    start_ptr, end_ptr = 0, 0
    
    result = 0
    num_rooms = 0
    
    length = len(intervals)
    
    #iterate until all meetings have been processed
    while start_ptr < length:
#         print(num_rooms)
        
        # if this meeting starts before the other ends, increment count and start ptr
        if start[start_ptr] < end[end_ptr]:
            start_ptr += 1
            num_rooms += 1
          
        # if this meeting ends before the other starts, decrement count (room is freed up) and increment end ptr
        else:
            num_rooms -= 1
            end_ptr += 1
            
        result = max(num_rooms, result)
            
    return result
            
        
    
    
    

In [28]:
intervals =[[13,15],[1,13],[6,9]]
minMeetingRooms(intervals)

0
1
2
1
0


2

<img src = "img/q3.png" style="width:600px;height:600px"/>

In [24]:
def insertInterval(intervals, newInterval):
    """
    :param intervals: Non-overlapping intervals sorted by start time
    :param newInterval: Start and end time of new interval that is to be inserted with or without merge
    :return: new list of intervals
    """
    output = []
    # iterate over the intervals
    for i in range(len(intervals)):

        # Non-overlapping case: end time of new interval is less than start time of current interval
        if (newInterval[1] < intervals[i][0]):
            # 1. insert the interval here without any merge
            output.append(newInterval)

            # 2. return this interval with the rest of the intervals as no merge possible in this case
            # output.extend(intervals[i:])
            return output + intervals[i:]
            # return output

        # Non-overlapping case: start time of current interval is more than the end time of current interval
        if (newInterval[0] > intervals[i][1]):
            # 1. insert the current interval to output
            output.append(intervals[i])

            # check the rest of the intervals before adding new interval

        # Overlapping case:
        else:
            # update the new Interval start time to be the min of start times of two intervals
            # update the new Interval end time to be the max of the two intervals
            newInterval[0] = min(intervals[i][0], newInterval[0])
            newInterval[1] = max(intervals[i][1], newInterval[1])

    # add the new Interval after iterating through all the intervals
    output.append(newInterval)

    return output

In [25]:
insertInterval([[4,5],[6,9],[8,10]],[1,3])

[[1, 3], [4, 5], [6, 9], [8, 10]]

In [26]:
insertInterval([[1,3],[6,9]],[2,5])

[[1, 5], [6, 9]]

In [27]:
insertInterval([[1,2],[3,5],[6,7],[8,10],[12,16]],[4,8])

[[1, 2], [3, 10], [12, 16]]

In [ ]:
def mergeIntervals(intervals):
    output = []
    # sort intervals by start time
    intervals.sort(key=lambda interval: interval[0])

    for i in range(len(intervals)-1):
        # non overlapping case
        if intervals[i][1] <intervals[i+1][0]:
            output.append(intervals[i])

        #overlapping case






<img src = "img/lc238.png" style="width:400px;height:500px"/>

In [13]:
def productExceptSelf(nums):
    length = len(nums)
    output = [1] * length
    multiplier = 1

    # get product of left side elems
    for i in range(length):
        output[i] *= multiplier
        multiplier *= nums[i]

    print("output left")
    print(output)

    # get product of right side elems
    multiplier = 1
    for i in range(length):
        # iterate in reverse order for right multiples
        output[length-1-i] *= multiplier
        multiplier *= nums[length-1-i]

    print("output total")
    print(output)

In [14]:
productExceptSelf([1,2,3,4])

output left
[1, 1, 2, 6]
output total
[24, 12, 8, 6]


## Type: Dynamic Programming (DP)

### LC 121: Best Time to Buy and Sell stock

Topics: Arrays, DP, Sliding Window

<img src = "img/q2.png" style="width:700px;height:600px"/>

In [15]:
def compute_best_time(prices:list[int]) -> int :

    # initialize mininum buying price, and maximum profit
    buy_price, profit = float("inf"), 0

    # iterate over the prices while the left pointer and right pointer don't meet
    for price in prices:
        # if this price is lower than buy price, update buying price
        buy_price = min(buy_price, price)

        # compute max of the diff between sell and buy price
        profit = max(profit, price-buy_price)

    return profit

In [16]:
prices = [7,1,5,3,6,4]
compute_best_time(prices)

5

In [17]:
prices = [7,6,4,3,1]
compute_best_time(prices)

0

<img src = "img/lc53.png"  style="width:600px;height:650px"/>

In [1]:
def getMaxSubArray(nums:list[int]) -> list[int]:

    # initialize start and end of the subarray window
    max_sum, current_max = -float("inf"), -float("inf")

    # iterate through the nums list
    for num in nums:
        current_max = max(num, current_max+num)
        max_sum = max(max_sum, current_max)

    return max_sum

In [2]:
getMaxSubArray([-2,1,-3,4,-1,2,1,-5,4])

6

In [19]:
getMaxSubArray([-2,1,-3,4,-1,2,1,2,-5,4])

8